src/payment_store.rs

```diff
@@ -0,0 +1,232 @@
+		let mut locked_payments = self.payments.lock().unwrap();
+
+		let payment_hash = payment_info.payment_hash.clone();
+		locked_payments.insert(payment_hash.clone(), payment_info.clone());
+
+		let key = format!(
+			"{}/{}",
+			PAYMENT_INFO_PERSISTENCE_PREFIX,
+			hex_utils::to_string(&payment_hash.0)
+		);
+
+		self.persister.persist(&key, &payment_info).map_err(|_| Error::PersistenceFailed)?;
+
+		return Ok(());
+	}
+
+	pub(crate) fn lock(&self) -> Result<PaymentInfoGuard<K>, ()> {
+		let locked_store = self.payments.lock().map_err(|_| ())?;
+		Ok(PaymentInfoGuard::new(locked_store, self.persister.clone()))
+	}
+
+	pub(crate) fn remove(&self, payment_hash: &PaymentHash) -> Result<(), Error> {
+		let key = format!(
+			"{}/{}",
+			PAYMENT_INFO_PERSISTENCE_PREFIX,
+			hex_utils::to_string(&payment_hash.0)
+		);
+		self.persister.unpersist(&key).map_err(|_| Error::PersistenceFailed)?;
+		Ok(())
+	}
+
+	pub(crate) fn get(&self, payment_hash: &PaymentHash) -> Option<PaymentInfo> {
+		self.payments.lock().unwrap().get(payment_hash).cloned()
+	}
+
+	pub(crate) fn contains(&self, payment_hash: &PaymentHash) -> bool {
+		self.payments.lock().unwrap().contains_key(payment_hash)
+	}
+
+	pub(crate) fn set_status(
+		&self, payment_hash: &PaymentHash, payment_status: PaymentStatus,
+	) -> Result<(), Error> {
+		let mut locked_payments = self.payments.lock().unwrap();
+
+		if let Some(p) = locked_payments.get_mut(payment_hash) {
+			p.status = payment_status;
+
+			let key = format!(
+				"{}/{}",
+				PAYMENT_INFO_PERSISTENCE_PREFIX,
+				hex_utils::to_string(&payment_hash.0)
+			);
+			self.persister.persist(&key, p).map_err(|_| Error::PersistenceFailed)?;
+		}
+		Ok(())
+	}
+}
+
+pub(crate) struct PaymentInfoGuard<'a, K: Deref>
+where
+	K::Target: KVStorePersister + KVStoreUnpersister,
+{
+	inner: MutexGuard<'a, HashMap<PaymentHash, PaymentInfo>>,
+	touched_keys: HashSet<PaymentHash>,
+	persister: K,
+}
+
+impl<'a, K: Deref> PaymentInfoGuard<'a, K>
+where
+	K::Target: KVStorePersister + KVStoreUnpersister,
+{
+	pub fn new(inner: MutexGuard<'a, HashMap<PaymentHash, PaymentInfo>>, persister: K) -> Self {
+		let touched_keys = HashSet::new();
+		Self { inner, touched_keys, persister }
+	}
+
+	pub fn entry(
+		&mut self, payment_hash: PaymentHash,
+	) -> hash_map::Entry<PaymentHash, PaymentInfo> {
+		self.touched_keys.insert(payment_hash);
+		self.inner.entry(payment_hash)
+	}
+}
+
+impl<'a, K: Deref> Drop for PaymentInfoGuard<'a, K>
+where
+	K::Target: KVStorePersister + KVStoreUnpersister,
+{
+	fn drop(&mut self) {
+		for key in self.touched_keys.iter() {
+			let store_key =
+				format!("{}/{}", PAYMENT_INFO_PERSISTENCE_PREFIX, hex_utils::to_string(&key.0));
+
+			match self.inner.entry(*key) {
+				hash_map::Entry::Vacant(_) => {
+					self.persister.unpersist(&store_key).expect("Persistence failed");
+				}
+				hash_map::Entry::Occupied(e) => {
+					self.persister.persist(&store_key, e.get()).expect("Persistence failed");
+				}
+			};
```

Comment:
So the drawback here is that we must crash if persistence fails. I suppose we already do that elsewhere when called from the event handler. More generally, is this what we should expect to do when persistence fails? We may need approach that gives a better UX.